components/Themes.js

```diff
@@ -0,0 +1,361 @@
+          display: flex;
+          flex: 1;
+          justify-content: ${item.id === 'create' ? 'center' : 'space-between'};
+          align-items: center;
+        }
+
+        .icon {
+          display: flex;
+        }
+      `}
+    </style>
+  </div>
+)
+
+const themeIcon = <ThemeIcon />
+
+class Themes extends React.PureComponent {
+  constructor(props) {
+    super(props)
+
+    const storedThemes = getThemes(localStorage) || []
+    const themes = [...storedThemes, ...THEMES]
+    const highlights = themes.find(({ id }) => id === props.theme).highlights
+
+    this.state = {
+      preset: props.theme,
+      themes,
+      highlights,
+      name: '',
+      selected: null
+    }
+  }
+
+  applyPreset = preset =>
+    this.setState(({ themes }) => ({
+      preset,
+      highlights: themes.find(({ id }) => id === preset).highlights
+    }))
+
+  handleDropdown = ({ id }) => {
+    if (id === 'create') {
+      this.props.toggleVisibility()
+    } else {
+      this.props.updateTheme(id)
+    }
+  }
+
+  updateName = ({ target: { value: name } }) => this.setState({ name })
+
+  selectHighlight = highlight => () =>
+    this.setState(({ selected }) => ({
+      selected: selected === highlight ? null : highlight
+    }))
+
+  updateHighlight = ({ rgb }) =>
+    this.setState({
+      highlights: {
+        ...this.state.highlights,
+        [this.state.selected]: stringifyRGBA(rgb)
+      }
+    })
+
+  removeTheme = id => () => {
+    const { themes } = this.state
+
+    const newThemes = themes.filter(t => t.id !== id)
+
+    saveThemes(localStorage, newThemes.filter(({ custom }) => custom))
+
+    if (this.props.theme === id) {
+      this.props.updateTheme(DEFAULT_THEME.id)
+    } else {
+      this.setState({ themes: newThemes })
+    }
+  }
+
+  createTheme = () => {
+    const { themes, name, highlights } = this.state
+
+    const id = generateId()
+
+    const newTheme = {
+      id,
+      name,
+      highlights,
+      custom: true
+    }
+
+    const customThemes = [newTheme, ...themes.filter(({ custom }) => custom)]
+
+    saveThemes(localStorage, customThemes)
+
+    this.props.updateTheme(id)
+  }
+
+  render() {
+    const { theme, isVisible, toggleVisibility } = this.props
+    const { name, themes, highlights, selected, preset } = this.state
+
+    const dropdownValue = isVisible
+      ? { name }
+      : { id: theme, name: themes.find(({ id }) => id === theme).name }
+
+    const dropdownList = [
+      {
+        id: 'create',
+        name: 'Create +'
+      },
+      ...themes
+    ]
+
+    return (
+      <div className="themes">
+        <Dropdown
+          icon={themeIcon}
+          disableInput={isVisible}
+          inputValue={dropdownValue}
+          selected={dropdownValue}
+          list={dropdownList}
+          itemWrapper={ThemeItem}
+          itemWrapperProps={{ onClick: this.removeTheme }}
+          onChange={this.handleDropdown}
+          onOpen={isVisible && toggleVisibility}
```

Comment:
So I need a callback to know when the dropdown is opened so that I can close this popout. Is there a better way for `Themes` to know when the dropdown is clicked open?

Reply:
Sounds good 👍